### DISCUSSION TITLE
What is the state of Mockery v3?

### DISCUSSION THREAD

**Discussion by anthony-hayes** [vektra/mockery#810] [General]
I noticed the v3 branch is "stale" and the last commit was 2 years ago. I'm wondering if I should migrate to the packages feature, and when roughly v3 is planned to release.

**Comment by LandonTClipp** [comment]
Good question. I do not have a specific date for v3. My philosophy on major version bumps is that it should solve an issue for the developers, the main issue being maintenance of deprecated features. I have noticed the community at large has opted to shift to `packages` and most of the tickets I get are for that, not the legacy config semantics. Therefore, I do not have a huge incentive to make the painful migration to v3. `packages` was designed to be exactly what v3 will be. The idea being that if you are using `packages` and you have no deprecation warnings, moving to v3 will not require any changes on your end beyond changing a single number in your installation steps.

There is another metric I go by. When enough config parameters have become deprecated, there comes a tipping point when it's time to do a broad cleanup of the codebase by removing the deprecated parameters entirely, which means a major version bump. That time has not arrived yet. Because of these reasons, I do not have any motivation to bump to v3 at this time, and I don't foresee it happening for at least a few more years. We are in this state primarily because the community has so enthusiastically adopted `packages` and I have not been focusing any energy at all into deprecated features.

**Reply by anthony-hayes** [reply]
That all makes a lot of sense. I'm glad to hear `packages` has been a success - it sounds like `packages` is the future for mockery regardless of the v3 release date. Thanks for the great response, this answers all my questions 🎉

**Comment by sbp-bvanb** [comment]
Regarding: `I do not have any motivation to bump to v3 at this time, and I don't foresee it happening for at least a few more years` Would it be an option then to remove the deprecation warning?

**Reply by LandonTClipp** [reply]
I don't think so. The old style of config is still deprecated because I don't want to support it anymore, and I do intend to remove it at some point.

**Comment by LandonTClipp** [comment]
Update on this, there is a new [Deprecations page](https://vektra.github.io/mockery/latest/deprecations/) on the docs. I'm starting to accumulate a number of deprecations at this point, so the balance is starting to shift towards a v3 release. Still no specific date yet but at this point the default config will spew a number of deprecation warnings, which is not a good user experience.

**Comment by LandonTClipp** [comment]
v3 has been released!
